cadc-access-control/src/main/java/ca/nrc/cadc/ac/client/GMSClient.java

```diff
@@ -1179,4 +1092,54 @@ protected RegistryClient getRegistryClient()
         return new RegistryClient();
     }
 
+    /**
+     * Lookup the Service URL for the given standard.  The current AuthMethod
+     * will be taken into account.
+     *
+     * @param standard  The URI standard to look up.
+     * @return          URL for the service.
+     * @throws AccessControlException       If the URL cannot be found for the
+     *                                      provided AuthMethod.
+     */
+    private URL lookupServiceURL(final URI standard)
+            throws AccessControlException
+    {
+        final URL serviceURL = getRegistryClient()
+                .getServiceURL(this.serviceID, standard, getAuthMethod());
+
+        if (serviceURL == null)
+        {
+            throw new AccessControlException(
+                    String.format("Unable to get Service URL for '%s', '%s', '%s'",
+                                  serviceID.toString(), Standards.GMS_GROUPS_01,
+                                  getAuthMethod()));
+        }
+        else
+        {
+            return serviceURL;
+        }
+    }
+
+    private AuthMethod getAuthMethod()
+    {
+        Subject subject = AuthenticationUtil.getCurrentSubject();
+        if (subject != null)
+        {
+            for (Object o : subject.getPublicCredentials())
+            {
+                if (o instanceof X509CertificateChain)
+                    return AuthMethod.CERT;
+                if (o instanceof SSOCookieCredential)
+                    return AuthMethod.COOKIE;
+                // AuthMethod.PASSWORD not supported
+                // AuthMethod.TOKEN not supported
+            }
+
+            throw new AccessControlException("No valid public credentials.");
+        }
+        else
+        {
```

Comment:
Since the GMSClient has knowledge of the capabilities of the service, it knows that it does not support anonymous access, so I would throw an AccessControlException here too (instead of returning ANON).